Request for help.

Current version of katagoWithCuda no longer builds.  Neither does this version.

The error received is during CMake's configure stage:

-- Check for working CUDA compiler: /nix/store/nzns9piy85q1x5hy43qgp5vny2qzwm1x-cudatoolkit-10.2.89/bin/nvcc
-- Check for working CUDA compiler: /nix/store/nzns9piy85q1x5hy43qgp5vny2qzwm1x-cudatoolkit-10.2.89/bin/nvcc - broken
CMake Error at /nix/store/fi94pv0wrajzx7bjxdabz1dk5nrf4dpb-cmake-3.17.0/share/cmake-3.17/Modules/CMakeTestCUDACompiler.cmake:46 (message):
  The CUDA compiler

    "/nix/store/nzns9piy85q1x5hy43qgp5vny2qzwm1x-cudatoolkit-10.2.89/bin/nvcc"

  is not able to compile a simple test program.

  It fails with the following output:

```
    Change Dir: /build/source/cpp/build/CMakeFiles/CMakeTmp

    Run Build Command(s):/nix/store/998ykzc5mazwak08hxayqmyqkyz18fcb-gnumake-4.2.1/bin/make cmTC_3083c/fast && /nix/store/998ykzc5mazwak08hxayqmyqkyz18fcb-gnumake-4.2.1/bin/make -f CMakeFiles/cmTC_3083c.dir/build.make CMakeFiles/cmTC_3083c.dir/build
    make[1]: Entering directory '/build/source/cpp/build/CMakeFiles/CMakeTmp'
    Building CUDA object CMakeFiles/cmTC_3083c.dir/main.cu.o
    /nix/store/nzns9piy85q1x5hy43qgp5vny2qzwm1x-cudatoolkit-10.2.89/bin/nvcc -forward-unknown-to-host-compiler   -cudart shared    -x cu -c /build/source/cpp/build/CMakeFiles/CMakeTmp/main.cu -o CMakeFiles/cmTC_3083c.dir/main.cu.o
    Linking CUDA executable cmTC_3083c
    /nix/store/fi94pv0wrajzx7bjxdabz1dk5nrf4dpb-cmake-3.17.0/bin/cmake -E cmake_link_script CMakeFiles/cmTC_3083c.dir/link.txt --verbose=1
    /nix/store/nr6f7jmhrxb4plz8n535s7fwymq0vns0-gcc-wrapper-7.5.0/bin/g++   CMakeFiles/cmTC_3083c.dir/main.cu.o -o cmTC_3083c  -lcudadevrt -lcudart_static  -L"/nix/store/nzns9piy85q1x5hy43qgp5vny2qzwm1x-cudatoolkit-10.2.89/targets/x86_64-linux/lib/stubs" -L"/nix/store/nzns9piy85q1x5hy43qgp5vny2qzwm1x-cudatoolkit-10.2.89/targets/x86_64-linux/lib" -lcudadevrt -lcudart
    /nix/store/3b3ighb83nhifa1v4n7855hlbdl1mhf9-binutils-2.31.1/bin/ld: /nix/store/8b71qygfrmbl9scwk83d5b589ghlhgbq-cudatoolkit-10.2.89-lib/lib/libcudart_static.a(libcudart_static.a.o): undefined reference to symbol 'dlclose@@GLIBC_2.2.5'
    /nix/store/3b3ighb83nhifa1v4n7855hlbdl1mhf9-binutils-2.31.1/bin/ld: /nix/store/bwzra330vib0ik4d3l8rq6gp6y2ah1fr-glibc-2.30/lib/libdl.so.2: error adding symbols: DSO missing from command line
    collect2: error: ld returned 1 exit status
    make[1]: *** [CMakeFiles/cmTC_3083c.dir/build.make:104: cmTC_3083c] Error 1
    make[1]: Leaving directory '/build/source/cpp/build/CMakeFiles/CMakeTmp'
    make: *** [Makefile:138: cmTC_3083c/fast] Error 2
```

I encountered this error last time around, and I remember it being a compiler version issue, which is why I was on gcc8Stdenv for the CUDA build.  This leads me to believe that there's perhaps an ABI mismatch between the new version of GCC (8.4.0) and the version that was available when this program was initially packaged (8.3.0).

Because ofBorg and hydra won't build Cuda stuff, this won't fail any of the automatic tests.

Soliciting ideas.

--- a/pkgs/games/katago/default.nix
+++ b/pkgs/games/katago/default.nix
@@ -6,6 +6,7 @@
 , cmake
 , makeWrapper
 , fetchFromGitHub
+, fetchpatch
 , cudnn ? null
 , cudatoolkit ? null
 , libGL_driver ? null
@@ -34,14 +35,29 @@ let
 
 in env.mkDerivation rec {
   pname = "katago";
-  version = "1.3.5";
+  version = "1.4.2";
 
   src = fetchFromGitHub {
     owner = "lightvector";
     repo = "katago";
     rev = "v${version}";
-    sha256 = "1625s3fh0xc2ldgyl6sjdjmpliyys7rzzkcys6h9x6k828g8n0lq";
+    sha256 = "0qdc9hgbzd175b2xkjs62dy6gyybcn9lf1mifiyhjbzjpgv192h4";
   };
+
+  # To workaround CMake 3.17.0's new buggy behavior wrt CUDA Compiler testing
+  # See the following tracking issues:
+  # KataGo:
+  #  - Issue #225: https://github.com/lightvector/KataGo/issues/225
+  #  - PR #227: https://github.com/lightvector/KataGo/pull/227
+  # CMake:
+  #  - Issue #20708: https://gitlab.kitware.com/cmake/cmake/-/issues/20708
+  patches = [
+    (fetchpatch {
+      name = "227.patch";
+      url = "https://patch-diff.githubusercontent.com/raw/lightvector/KataGo/pull/227.patch";
+      sha256 = "03f1vmdjhb79mpj95sijcwla8acy32clrjgrn4xqw5h90zdgj511";
+    })
+  ];
 
   nativeBuildInputs = [
     cmake

